### ISSUE TITLE
Let's sort GHCJS boot

### ISSUE THREAD

**Issue by tolysz** [ghcjs/ghcjs#448]
As there is a lot of new packages; people wanting to bootstrap using custom packages; or ones from some preselected collection.

Now the wild idea is to have a patching/cache proxy; possibly some local app.
which would:

    request somepackage-1.2.4

    request p = 
       response =<< check cache >>= \case
           Just a ->  return a
           Nothing -> download from upstram >>=\case
                 OK a -> return $ store p (patch p a)
                 _ -> error

 patch would need to repack the content; (optionally add patch version to the package version)

This will require having collections of patches for different versions; currently it would be useful as we could have `base-4.8.0.0` `base-4.8.1.0` `base-4.8.2.0` and all other from boot; 

Where to store them? I would prefer `github` or some ghcjs shadow layer in hackage.

So the result: we could remove boot and have everything patched on the fly (we might need to have ghcjs specific packages pre-populating cache)

**Comment by mgsloan** [comment]
I agree that it would be great to support patching other versions of the packages (particularly to support the versions used by stackage snapshots).

It'd be interesting to see how much mileage we can get out of the approach you suggest, but I'm guessing that many of the patches will fail to apply to different versions of the code.  There also should be some manual aspect of the process, as features may have been added to the library that also need to be patched / shimmed.

I think the ideal solution would be to put the GHCJS patches directly into the libraries.  I don't think this would be necessary for wired-in packages, though, because .

If that doesn't work out, we're going to need some way to release GHCJS specific versions of packages.  I don't see the `ghcjs-boot` style of patching packages being very sustainable in the long run.  Perhaps we just need some cabal or hackage metadata that says "The GHCJS version of `process` this package is called `ghcjs-process`".  This wouldn't resolve the maintenance burden of uploading GHCJS versions of these packages, though.

Misc question: Why does ghcjs-boot install aeson and similar libraries?  This is one of the main things that's causing strife for people using ghcjs with stackage snapshots.

**Comment by luite** [comment]
I don't think we can really get rid of `ghcjs-boot` yet, since the compiler is only fully functional when a base set of packages, including `Cabal` has been installed. This is the same for GHC, and solving this would probably be done there. We also need the `ghcjs-prim` wired-in package in addition to those that GHC needs (otherwise the compiler can't run template haskell and some symbols that the RTS depends on are missing)

However, we don't really need to install as many packages with `ghcjs-boot` as we do now, things like `aeson` and `text` are dependencies of `ghcjs-base`, which is actually not a special package. It can be installed with cabal without problems (although version compatibility between GHCJS versions is rather limited due to the large amount of low-level code). The reason that these things are still built by `ghcjs-boot` really is those patches in the packages.

I think the best long-term solution is to merge GHCJS support patches upstream. Say we start doing this for GHC(JS) 8.0 with base 4.9.x, then all known-to-fail packages can get this in a metadata update:

```
if impl(ghcjs) build-depends: ghcjs_unsupported<0
```

which prevents selection by the solver. GHCJS fixes can of course be backported to earlier versions. Packages that don't build with base 4.9 don't need to be updated.

Thanks to @hvr for this idea. Perhaps we can do better by modifying Cabal to more directly express supported and unsupported build environments though? Ideas welcome.

**Comment by tolysz** [comment]
So if we could drop to only essential packages. But then we still need to have `on-the-flay` patching
for all packages. I thought patches will be independent of the `ghcjs` version; but it would prevent @luite to change internals. So we need to have some way having possible different patches for:
 `ghcjs-version_ghc-version_package-version`
we already have it for new and old base...

My goal is not to "cancel" `ghcjs-boot` but to be able to easily build a new collection of packages.
and be able to track those patches more clearly.

Until, the maintainers will be accept multiple `#CPP`s for different versions of `ghcjs` we need to have an easy way of maintaining the collection of patches (and shims).

I am not sure what will be the minimal approach here.
I would however start with:

0. given: `ghcjs-version`; `ghc-version`; `package-version`
1. download: package and apply patches
2. a. be able to build new boot.tar from a list of packages
2. b. be able to boot `ghcjs` using list
3. have `cabal` or `stack` be able to patch packages using some service.
4. and use them for installation

**Comment by mgsloan** [comment]
Hopefully for stack users, things will be better here after this commit -https://github.com/commercialhaskell/stack/pull/1514 .  @luite Seems like "stage 1" packages are all the ones that need to be patched (or are deps of the patched packages).  Does that seem correct?

**Comment by tolysz** [comment]
I think `aeson` is needed for `ghcjs-base` which is not a published package; thus there will be no way to install it later.

Unfortunately ANY package might need to be patched, but I think there are some heuristics on what would need to be... my guess anything with `#if defined (__MINGW32__)`

**Comment by Ericson2314** [comment]
Say we just fork the necessarily upstream packages and apply the patches, and use a `stack.yaml` to lock the necessary versions. It isn't the most flexible, but I think that's solely better than the status quo.

Long term, putting the proper CPP upstream is still an option---the transition is seemless, due the `ghcjs-unsupported` trick, normal cabal constraints, and switching `stack.yaml` to use hackage instead of git.

As far as I can tell, with that in mind we can get rid of ghcjs-boot right now.

(I earlier brought up Stack in https://github.com/ghcjs/ghcjs/issues/442)

**Comment by tolysz** [comment]
I have been looking into that for a while, there is only a few packages needed for boot.
And only a very few requing patching. (e.g. https://hackage.haskell.org/package/base )
I followed apporach; of getting current versions from the stackage to build the boot.
The approach: get versions from resolver; get packages needed for boot; patch them; build boot install; use different `ghcjs` boot directories (i.e. include the resolver in the path)

My sugestion would be to nominate someone as "release manager" for `ghc 7.10` `0.2`... and give @luite room for `ghc-8` and possibly `0.3`...

Current effort will be to update `base` i.e. try to port changes.
ps.
(I am stalled as `base-4.8.2.0` is not on hackage - pls )

**Comment by Ericson2314** [comment]
Cross post from #442 I've started stackifying things here https://github.com/forked-upstream-packages-for-ghcjs

**Comment by tolysz** [comment]
Hi, https://github.com/tolysz/ghcjs-stack is an idea how to align packages from `ghcjs-boot` with packages for a particular resolver with stack. It is hard to have "one to boot them all" solution.

Currently, it looks like porting patches for `integer-gmp` to version `1.0.0.0` will be the biggest effort(is anyone working on that? )

Using the above repo's scripts one can "create" a version for any particular resolver; but we might need to have a DB with patches, for now a folder is just fine.

**Comment by Ericson2314** [comment]
https://github.com/ghcjs/ghcjs/issues/467
